src/polkit/polkitunixprocess.c

```diff
@@ -1131,6 +1361,437 @@ get_start_time_for_pid (pid_t    pid,
   return start_time;
 }
 
+static guint64
+get_cgroupid_for_pidfd (gint    pidfd,
+                        GError **error)
+{
+#ifdef __linux__
+  guint64 cgroupid;
+  gint pid;
+  gchar *filename;
+  gchar *cgroup_path;
+  gchar *contents;
+  size_t length;
+  gchar *p;
+  struct file_handle *fhp;
+  int mount_id;
+
+  pid = 0;
+  filename = NULL;
+  contents = NULL;
+  cgroup_path = NULL;
+  cgroupid = 0;
+  fhp = NULL;
+
+  g_return_val_if_fail (pidfd >= 0, 0);
+
+  pid = polkit_unix_process_get_pid_from_pidfd (pidfd, error);
+  if (pid <= 0)
+    return 0;
+
+  filename = g_strdup_printf ("/proc/%d/cgroup", pid);
```

Comment:
this is ugly and racy. maybe just use the new ioctl to get the cgroup id directly from the pid?

Reply:
because cgroups can be removed/recreated, and are different beasts then, as per cgroup id. if you go directly from pid to cgroupid you are not vulnerable to races around that hence.